### PR TITLE
ref(snuba): Refactor utils/snuba to allow for multiple referrers

### PR DESCRIPTION
**Pull request by Zylphrex**
Snuba queries only allow for a single referrer for a batch of parallel queries. But sometimes, these queries have nothing to do with each other and using the same referrer is not ideal. This refactor prepares the utilities to allow callers to pass a referrer per query.

### CODE REVIEW

**Review by mjq — APPROVED** [review]
Nice!

File: src/sentry/utils/snuba.py [hunk 1]
```diff
@@ -847,9 +848,31 @@ def raw_query(
...
 
-SnubaQuery = Union[Request, MutableMapping[str, Any]]
 Translator = Callable[[Any], Any]
-RequestQueryBody = tuple[Request, Translator, Translator]
+
+
+@dataclasses.dataclass(frozen=True)
+class SnubaRequest:
+    request: Request
+    referrer: str | None  # this should use the referrer class
```
Comment (wmak) [inline src/sentry/utils/snuba.py:857]: Is this a TODO?

File: src/sentry/utils/snuba.py [hunk 2]
```diff
@@ -847,9 +848,31 @@ def raw_query(
...
+    forward: Translator
+    reverse: Translator
+
+    def __post_init__(self) -> None:
+        self.validate()
+
+    def validate(self):
+        if self.referrer:
```
Comment (wmak) [inline src/sentry/utils/snuba.py:865]: Huh, this matches previous behaviour but it feels so weird that referrer isn't required

File: tests/sentry/issues/test_attributes.py [hunk 3]
```diff
@@ -237,14 +235,7 @@ def run_attr_test(
...
             request.validate()
-            identity = lambda x: x
-            isolation_scope = sentry_sdk.Scope.get_isolation_scope().fork()
-            current_scope = sentry_sdk.Scope.get_current_scope().fork()
-            resp = _snuba_query(
-                ((request, identity, identity), isolation_scope, current_scope, {}, "test_api")
-            )
-            assert resp[0].status == 200
-            stuff = json.loads(resp[0].data)
+            stuff = raw_snql_query(request)
```
Comment (wmak) [inline tests/sentry/issues/test_attributes.py:238]: ... `stuff`, I know you're reusing the existing variable names but what the heck

**Review by wmak — APPROVED** [review]
Small comments, lgtm

File: src/sentry/utils/snuba.py (hunk 1 above)
Comment (Zylphrex) [reply src/sentry/utils/snuba.py:857]: Yes, I'll prefix the comment with a TODO once I figure out if we can change this in all the call sites.

File: src/sentry/utils/snuba.py (hunk 2 above)
Comment (Zylphrex) [reply src/sentry/utils/snuba.py:865]: Yeah, I thought this was weird too but let's leave the behaviour unchanged in this PR

File: tests/sentry/issues/test_attributes.py (hunk 3 above)
Comment (Zylphrex) [reply tests/sentry/issues/test_attributes.py:238]: Good point, let me rename all these to something more meaningful since I'm here anyways

**Pull request merged**